torch/__init__.py

```diff
@@ -616,37 +616,45 @@ def _assert(condition, message):
 # Import most common subpackages
 ################################################################################
 
-import torch.cuda
-import torch.autograd
-from torch.autograd import no_grad, enable_grad, set_grad_enabled
-import torch.fft
-import torch.futures
-import torch.nn
+# Use the redundant form so that type checkers know that these are a part of
+# the public API. The "regular" import lines are there solely for the runtime
+# side effect of adding to the imported module's members for other users.
+
+from torch import cuda as cuda
+from torch import autograd as autograd
+from torch.autograd import (
+    no_grad as no_grad,
+    enable_grad as enable_grad,
+    set_grad_enabled as set_grad_enabled,
+)
+from torch import fft as fft
+from torch import futures as futures
+from torch import nn as nn
 import torch.nn.intrinsic
 import torch.nn.quantizable
 import torch.nn.quantized
-import torch.optim
+from torch import optim as optim
 import torch.optim._multi_tensor
-import torch.multiprocessing
-import torch.sparse
+from torch import multiprocessing as multiprocessing
+from torch import sparse as sparse
 import torch.utils.backcompat
-import torch.onnx
-import torch.jit
-import torch.linalg
-import torch.hub
-import torch.random
-import torch.distributions
-import torch.testing
+from torch import onnx as onnx
+from torch import jit as jit
+from torch import linalg as linalg
+from torch import hub as hub
+from torch import random as random
+from torch import distributions as distributions
+from torch import testing as testing
```

Comment:
I'm not entirely certain this is in the public API, but I can see many places where it's used as `torch.testing` without `import torch.testing`, and a few where it is.

Reply:
can you describe the rules you used in this PR?  Like, it's top level namespaces plus things that were already "from ... import ..." and "import ... as ..." in this file?

Reply:
This block is all things that were unused by any other part of the file, and the above header implied that their existence was solely to ensure that these modules were loaded and available to anyone who had just done `import torch`, so I converted them. Others, I followed what is exposed in the official documentation as is here: https://github.com/pytorch/pytorch/issues/47027#issuecomment-761851536

I have not pushed down these other side effects either; the other unused imports should eventually also move (if possible), but that is a semantic change.

Reply:
Effectively, if something should "be a part of the torch API", but comes from another module, it either needs the redundant form or explicit assignment.